File: go/prueba.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func helloThread(initialNumber, finalNumber int, wg *sync.WaitGroup) {
	defer wg.Done()

	for i := initialNumber; i < finalNumber; i++ {
		fmt.Printf("%d-\n", i)
	}
}

func count(numberThreads, number int) {
	base := number / numberThreads
	restante := number % numberThreads

	inicial := 0

	var wg sync.WaitGroup

	start := time.Now()

	for i := 0; i < numberThreads; i++ {

		fin := inicial + base

		if i < restante {
			fin++
		}

		wg.Add(1)

		go helloThread(inicial, fin, &wg)

		inicial = fin
	}

	// Esperar a que todas las goroutines terminen
	wg.Wait()

	elapsed := time.Since(start)

	fmt.Printf("\nTiempo total: %d ns\n", elapsed.Nanoseconds())
}

func main() {

	var total int
	var threads int

	fmt.Print("Ingresa el número total: ")
	fmt.Scan(&total)

	fmt.Print("Ingresa la cantidad de threads: ")
	fmt.Scan(&threads)

	if threads <= 0 {
		fmt.Println("La cantidad de threads debe ser mayor que 0.")
		return
	}

	if total < 0 {
		fmt.Println("El número total debe ser positivo.")
		return
	}

	count(threads, total)
}

File: javaX/Prueba.java

package javaX;

import java.util.Scanner;

public class Prueba {

    public void count(int numberThreads, int number){
        int porcion = number / numberThreads;
        int inicial = 0;
        int fin = porcion; 
        long inicio = System.nanoTime();
        for(int i=0; i < numberThreads; i ++){
            HelloThread thread = new HelloThread(inicial,fin);
            thread.start();
            inicial = (i+1)*porcion;
            fin = fin + porcion; 
        } 
        long fina = System.nanoTime();
        System.out.println("Tiempo: " + (fina - inicio) + " ns");
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in); // Crear scanner
        System.out.print("Ingresa el número total: ");
        int total = sc.nextInt();
        System.out.print("Ingresa la cantidad de threads: ");
        int threads = sc.nextInt();
        
        Prueba p = new Prueba();
        p.count(threads, total);
        sc.close();
    }
}